### ISSUE TITLE
Details for reproducing the indoor pose estimation experiment

### ISSUE THREAD

**Issue by angshine** [magicleap/SuperGluePretrainedNetwork#31]
Hi, thanks for the great work!

I'm reproducing SuperGlue on the ScanNet dataset and having some questions about the details of data preprocessing and ground-truth generation.
1. About data preprocessing. It is mentioned in the appendix-E that you select image pairs that have an overlap score in [0.4, 0.8], however, it is time-consuming to exhaustively search all possible image pairs and calculate the overlap score. Did you use any heuristic to prune out part of the pairs before calculating the overlap score? I think this is not trivial since different implementation might lead to a huge difference in the portion of pairs with a small overlap score thus causing different portions of hard training samples for the network. Furthermore, could you share with me what depth threshold is used when checking depth consistency?
2. About determining the ground-truth correspondences. It is mentioned in appendix-E that correspondences are determined by the reprojection error. After extracting the correspondences with minimum reprojection error along rows and columns with a threshold, did you check their depth consistency, if so,  what threshold is used?

**Comment by Skydes** [comment]
Hi,
Great to see reproduction efforts :)

1. Yes computing the overlap is expensive. To reduce the load, you can:
   1. Subsample pixels - I think that we used 1 pixel every 10 in both x and y directions
   2. Subsample sequences by keeping every N images - we did not do this, but 1/5 seems reasonable.
   3. Discard images with non-overlaping frustums, e.g. back-to-back cameras - we did not do this either but that would help a lot.

  These three heuristics should not impact much the overlap scores. Regarding depth threshold, we used 10% relative error.
  
2. Yes the depth has to be consistent, and we accept 5% relative error for ScanNet.

**Comment by angshine** [comment]
@Skydes Thanks for the reply and suggestions! :) Following your instruction, my re-implementation got a slightly worse result than the pre-trained model published (indoor model):
| model  | AUC@5 | AUC@10 | AUC@20 | Precision | MatchingScore |
| --------- | ----- | ------ | ------ | --------- | ------------- |
| SuperGluePretrained (bs=64, matching-thresh=0.2)  | 16.23 | 34.12  | 52.88  | 84.33     | 31.45         |
| re-implementation (bs=192, matching-thresh=0.2)| 14.90 | 32.07|50.13|84.52|29.02|
| re-implementation (bs=192, matching-thresh=0.05) | 16.67  | 33.45  | 51.59     | 82.61   | 31.58|

In my re-implementation, I use a larger batch-size and scale the learning schedule linearly w.r.t batch-size and replace the constant warm-up with a linear gradual warm-up. The dataset shares the same size as you mentioned in the paper - 230M.  I suspect that performance degeneracy is caused by the following aspect:
1. The increased batch size affects the model's generalization.
2. We might use a different initial value for the dustbin. I use 1.0 as initialization and find 0.0 got a worse result. Do you have any suggestions on the choices of the initial dustbin value for ScanNet?  I think the trained dustbin value of the pre-trained model is about 2.5, so maybe a larger dustbin value is preferred for ScanNet?
3. It was mentioned in the paper that SuperGlue is trained on ScanNet for a total of 900K iterations, 200 samples per scene are randomly sampled at each epoch, which means that the model gets trained on about 57M image pairs in total, hence the full dataset is not entirely covered. So the randomness of data sampling might have an unignorable effect on the model's final performance.
4. Did you use any data augmentation for input images of SuperPoint during training to augment the input descriptors for SuperGlue?
5. Did you balance the positive and negative loss terms, or just keep the loss function exactly the same as in the paper?

**Comment by Skydes** [comment]
Great job! Looking at the numbers, the matching score is a bit lower than our model, which hints a lack of supervision of the positive matches. Do you monitor the number of matched vs ummatched points used during training? As said in https://github.com/magicleap/SuperGluePretrainedNetwork/issues/24#issuecomment-670723004, if there is a large imbalance between positive and negative matches then balancing the loss can be necessary. We used equal (50%) balancing of positives and negatives for ScanNet - I admit that this should have been detailed in the paper. We did not use any data augmentation as the SuperPoint descriptors are already quite robust, but this could indeed help to generalize to other datasets.

I don't think that the large batch size can degrade the performance - it usually improves it. What is the improvement of you implementation between bs=64 and bs=192?

**Comment by angshine** [comment]
I think you are right，the decrease of matching scores is caused by lack of supervision of positive matches. For the above results, I assign equal weights to the three terms of the the loss function, which is exactly the same as the superglue paper. I have tried to assign more weights on the positive term, or augment the original loss with a focal loss, which leads to higher matching scores on my validation set. But the training has not been done, I will share with you new results later. I also find that a higher initial dustbin value leads to a better result on ScanNet(at least for my validation set while training), which seems counterintuitive.

I haven't done any experiments with bs=64, since it needs more time training :)

**Comment by Cuistiano** [comment]
> Hi,
> Great to see reproduction efforts :)
> 
> 1. Yes computing the overlap is expensive. To reduce the load, you can:
>    
>    1. Subsample pixels - I think that we used 1 pixel every 10 in both x and y directions
>    2. Subsample sequences by keeping every N images - we did not do this, but 1/5 seems reasonable.
>    3. Discard images with non-overlaping frustums, e.g. back-to-back cameras - we did not do this either but that would help a lot.
> 
> These three heuristics should not impact much the overlap scores. Regarding depth threshold, we used 10% relative error.
> 
> 1. Yes the depth has to be consistent, and we accept 10% relative error.

Hello, have you used depth consistency check when preprocessing outdoor dataset (Megadepth)? If so, the threshold is still 10% or another?

**Comment by Skydes** [comment]
We used the same 10% relative depth error for MegaDepth as well.
